backend/src/main/java/io/papermc/hangar/service/api/VersionsApiService.java

```diff
@@ -158,62 +151,71 @@ private boolean match(final Set<String> toAdd, final String selectedVersion, fin
         }
 
         final Long userId = this.getHangarUserId();
-        Stream<Version> versions = this.versionsApiDAO.getVersions(projectTable.getSlug(), canSeeHidden, userId, pagination).entrySet().parallelStream()
-            .map(entry -> this.versionDependencyService.addDownloadsAndDependencies(projectTable.getOwnerName(), projectTable.getSlug(), entry.getValue().getName(), entry.getKey()).applyTo(entry.getValue()))
+        Stream<Version> versions = this.versionsApiDAO.getVersions(project.getSlug(), canSeeHidden, userId, pagination).entrySet().parallelStream()
+            .map(entry -> this.versionDependencyService.addDownloadsAndDependencies(project.getOwnerName(), project.getSlug(), entry.getValue().getName(), entry.getKey()).applyTo(entry.getValue()))
             .sorted((v1, v2) -> v2.getCreatedAt().compareTo(v1.getCreatedAt()));
 
         if (!includeHiddenChannels) {
             versions = versions.filter(version -> !version.getChannel().getFlags().contains(ChannelFlag.HIDE_BY_DEFAULT));
         }
 
-        final Long versionCount = this.versionsApiDAO.getVersionCount(projectTable.getSlug(), canSeeHidden, userId, pagination);
+        final Long versionCount = this.versionsApiDAO.getVersionCount(project.getSlug(), canSeeHidden, userId, pagination);
         return new PaginatedResult<>(new Pagination(versionCount == null ? 0 : versionCount, pagination), versions.toList());
     }
 
-    public Map<String, VersionStats> getVersionStats(final String slug, final String versionString, final OffsetDateTime fromDate, final OffsetDateTime toDate) {
+    public Map<String, VersionStats> getVersionStats(final ProjectVersionTable version, final OffsetDateTime fromDate, final OffsetDateTime toDate) {
         if (fromDate.isAfter(toDate)) {
             throw new HangarApiException(HttpStatus.BAD_REQUEST, "From date is after to date");
         }
-        return this.versionsApiDAO.getVersionStats(slug, versionString, fromDate, toDate);
+        return this.versionsApiDAO.getVersionStats(version.getProjectId(), version.getId(), fromDate, toDate);
     }
 
     @CacheEvict(value = CacheConfig.LATEST_VERSION, keyGenerator = "ignoringCaseCacheKeyGenerator")
     public void evictLatestRelease(final String slug) {
     }
 
     @Cacheable(value = CacheConfig.LATEST_VERSION, keyGenerator = "ignoringCaseCacheKeyGenerator")
-    public @Nullable String latestVersion(final String slug) {
-        return this.latestVersion(slug, this.config.channels.nameDefault());
+    public @Nullable String latestVersion(final ProjectTable project) {
+        return this.latestVersion(project, this.config.channels.nameDefault());
     }
 
     @CacheEvict(value = CacheConfig.LATEST_VERSION, keyGenerator = "ignoringCaseCacheKeyGenerator")
     public void evictLatest(final String slug, final String channel) {
     }
 
     @Cacheable(value = CacheConfig.LATEST_VERSION, keyGenerator = "ignoringCaseCacheKeyGenerator")
```

Comment:
I think these cacheables/cacheevict need more attention. you prolly didnt see that I added them, but right now I don't think the eviction would work since the key is different. the evict methods should just have a long param for the slug I think and the cacheable methods need a key = "#project.getId(), #channel" or something (not sure how the syntax is)

Reply:
Setting both key and keyGenerator is not possible. Should I create another key generator or should I just add an instanceof check for the Identified (or perhaps Table?) class and use the id in the existing key generator?

Reply:
I think you can call #channel.toLowerCase() or something in SpEL? so we dont need the generator then